README.pod
=pod

=begin html

<a href="https://travis-ci.org/ivanwills/App-VTide?branch=master"><img src="https://travis-ci.org/ivanwills/App-VTide.png?branch=master" alt="App-VTide"></a>
<a href="https://coveralls.io/r/ivanwills/App-VTide?branch=master"><img src="https://coveralls.io/repos/ivanwills/App-VTide/badge.svg?branch=master" alt="Coverage Status" /></a>

=end html

=head1 App::VTide

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

=head1 INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

=head1 SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::VTide

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-VTide

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-VTide

    CPAN Ratings
        http://cpanratings.perl.org/d/App-VTide

    Search CPAN
        http://search.cpan.org/dist/App-VTide/

    Source Code
        git://github.com/ivanwills/App-VTide.git

=head1 COPYRIGHT AND LICENCE

Copyright (C) 2016 Ivan Wills

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut


Changes
Revision history for App-VTide

1.0.3   2023-06-25
        Updating for release of version 1.0.3
        Making default local file be more environment safe (Ivan Wills)
        Trying alternate title setting (Ivan Wills)
        New sessions default config (Ivan Wills)
        Adding saving current session (Ivan Wills)
        Adding missing param docs (Ivan Wills)
        Adding missing param docs (Ivan Wills)
        Fixing incorrect versions (Ivan Wills)

1.0.2   2023-05-31
        Updating for release of version 1.0.2
        Updating the manifest (Ivan Wills)
        Adding missing tests (Ivan Wills)

1.0.1   2023-04-01
        Updating for release of version 1.0.1
        Adding missing new files (Ivan Wills)
        Switching to using opt rather than default (Ivan Wills)
        Fixing list logic (Ivan Wills)
        Fixing auto completion and making list clearer (Ivan Wills)
        Fixing some ambiguities with auto completion (Ivan Wills)
        Minor fixes (Ivan Wills)
        New session management command (Ivan Wills)
        Fixing up adding sessions (Ivan Wills)
        Starting to add logic to save/restore sessions (Ivan Wills)
        Allowing recurse to work when a helper is defined (Ivan Wills)

0.1.21   2023-03-14
        Updating for release of version 0.1.21
        Updating documentation of commands (Ivan Wills)

0.1.20   2023-02-13
        Updating for release of version 0.1.20
        Updated (Ivan Wills)
        Trying to make starting more consistent on a Mac (Ivan Wills)
        Better history and recursive edits (Ivan Wills)

0.1.19   2022-12-04
        Updating for release of version 0.1.19
        Adding missing tests (Ivan Wills)
        Ignoring require should not use strings (it must in these cases (Ivan Wills)
        Fixing expected spellings errors (Ivan Wills)
        Returned accedentally deleted line (Ivan Wills)
        Adding helper new commands (Ivan Wills)
        Adding logging of run errors (Ivan Wills)
        Fixing duplicate files (Ivan Wills)
        Making this more mac compatible (Ivan Wills)
        Removing File::Touch (Ivan Wills)

0.1.18   2022-10-09
        Updating for release of version 0.1.18
        Adding the ability to search or exclude from history (Ivan Wills)

0.1.17   2022-07-03
        Updating for release of version 0.1.17
        Updated (Ivan Wills)
        Showing other custom environment variables (Ivan Wills)
        Removing odd return (Ivan Wills)
        Fixing help to be more helpful (Ivan Wills)
        Adding support for 1 level parent configs (Ivan Wills)
        Trying harder to set title (Ivan Wills)
        Adding logging exit times (Ivan Wills)

0.1.16   2021-09-05
        Updating for release of version 0.1.16
        Future note (Ivan Wills)
        Fixing the tmux title (Ivan Wills)
        Better quoting (Ivan Wills)
        Adding documentation for new cron config (Ivan Wills)
        Adding new cron helper that will run the terminal's command via a cron config (Ivan Wills)

0.1.15   2021-03-20
        Updating for release of version 0.1.15
        Fixing issue with undefined value (Ivan Wills)

0.1.14   2021-01-04
        Updating for release of version 0.1.14
        Updated date (Ivan Wills)
        Adding show option to restart menu (Ivan Wills)

0.1.13   2020-07-01
        Updating for release of version 0.1.13
        Updaging docs (Ivan Wills)
        Adding the edited files to the title (Ivan Wills)
        Adding set option that gives you the environment variables to set based on a vtide config (Ivan Wills)

0.1.12   2020-06-02
        Updating for release of version 0.1.12
        Making sure check is as accurate as possible (Ivan Wills)
        Making all commands use standard param for sizes (Ivan Wills)
        Trying to fix new terminal issue (Ivan Wills)
        Fixed history display (Ivan Wills)
        Made history (better) (Ivan Wills)

0.1.11   2020-05-19
        Updating for release of version 0.1.11
        Updated date (Ivan Wills)
        Adding missing files for testing (Ivan Wills)
        Added missing dependency (Ivan Wills)
        Moving run text to after the clear screen and printing only on the first run (Ivan Wills)
        Fixing case when watching files and trying to quit (Ivan Wills)
        Making sub-command config dynamic (Ivan Wills)
        Cleaning up history (Ivan Wills)
        New history command (Ivan Wills)
        Added more logic to help helper functions from having issues (Ivan Wills)
        Adding option to clear screen before re-running command (Ivan Wills)
        Fixing up travis build versions (Ivan Wills)
        Adding option to clear terminals before running (Ivan Wills)

0.1.10   2019-04-30
        Updating for release of version 0.1.10
        Adding missing recommends section (Ivan Wills)

0.1.9   2019-04-02
        Updating for release of version 0.1.9
        Split passing the title or name to pass both to stop code confusion (Ivan Wills)

0.1.8   2019-03-04
        Updating for release of version 0.1.8
        Adding new manifest files to skip (Ivan Wills)
        Better code comment (Ivan Wills)
        Changed recent to list newest first (Ivan Wills)
        Fixed issue when no arguments passed (Ivan Wills)
        Fixing up how autocompletion (Ivan Wills)
        Setting the minimum required version of Getopt::Alt to 0.5.2 (Ivan Wills)
        Fixing the way sub-command if found for auto-complete (Ivan Wills)
        Adding logs about command run (Ivan Wills)

0.1.7   2019-01-06
        Updating for release of version 0.1.7
        Fixning config dependencies (Ivan Wills)
        Updated date (Ivan Wills)
        Added getting absolute path for Perl 5.26 (Ivan Wills)

0.1.6   2018-03-07
        Updating for release of version 0.1.6
        Added perl 5.26 to travis builds (Ivan Wills)
        Updated license date (Ivan Wills)
        Cleaning up skip file (Ivan Wills)

lib/App/VTide/Command/Run.pm
package App::VTide::Command::Run;

# Created on: 2016-01-30 15:06:40
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo;
use warnings;
use version;
use Carp                qw/carp longmess/;
use English             qw/ -no_match_vars /;
use Hash::Merge::Simple qw/ merge /;
use Path::Tiny;
use File::stat;
use File::chdir;
use IO::Prompt qw/prompt/;
use Algorithm::Cron;
use List::MoreUtils qw/uniq/;
use Data::Dumper    qw/Dumper/;

extends 'App::VTide::Command';

our $VERSION = version->new('1.0.3');
our $NAME    = 'run';
our $OPTIONS = [ 'name|n=s', 'test|T!', 'save|s=s', 'verbose|v+', ];
our $LOCAL   = 1;
sub details_sub { return ( $NAME, $OPTIONS, $LOCAL ) }

has first => (
    is      => 'rw',
    default => 1,
);

has base => ( is => 'rw', default => $CWD );

sub run {
    my ($self) = @_;

    my ($name) = $self->session_dir( $self->defaults->{name} );
    my $cmd = $self->options->files->[0] || '';
    $ENV{VTIDE_TERM} = $cmd;

    my $params = $self->params($cmd);
    my @cmd    = $self->command($params);
    $self->log( 'START', @cmd );

    @ARGV = ();
    if (
        !(
               $self->first
            && ( $params->{watch} || $params->{cron} )
            && $params->{wait}
        )
      )
    {

        if ( $params->{clear} ) {
            system 'clear';
        }
        if ( $self->first ) {
            print "Running $name - $cmd\n";
        }

        if ( $params->{heading} ) {

            # show terminal heading if desired
            print $params->{heading}, "\n";
        }

        if ( !$self->defaults->{test} && $params->{wait} ) {
            print join ' ', @cmd, "\n";
            print "Press enter to start : ";
            my $ans = <ARGV>;
            if ( !$ans || !ord $ans ) {
                print "\n";
                return;
            }
        }

        $self->load_env( $params->{env} );
        local $CWD = $CWD;
        $self->base($CWD);
        if ( $params->{dir} && -d $params->{dir} ) {
            $CWD = $params->{dir};
        }

        if ( $self->defaults->{verbose} || $self->defaults->{test} ) {
            warn "Will wait before starting\n" if $params->{wait};
            warn "Will restart on exit\n"      if $params->{restart};
        }

        # run any hooks for run_running
        $self->hooks->run( 'run_running', \@cmd );

        # start the terminal
        eval {
            $self->runit(@cmd);
            1;
        } or do {
            $self->log( "RUN ERROR", @cmd, $@ );
        }
    }

    # flag this is no longer the first run
    $self->first(0);

    if ( !$self->defaults->{test} && $self->restart($cmd) ) {
        return $self->run;
    }

    return;
}

sub restart {
    my ( $self, $cmd, $no_watch ) = @_;

    my $params = $self->params($cmd);

    return $self->watch($cmd) if !$no_watch && $params->{watch};
    return $self->cron($cmd)  if !$no_watch && $params->{cron};

    return if !$params->{restart};

    my %action = (
        q => {
            msg  => 'quit',
            exec => sub { 0; },
        },
        c => {
            msg  => 'clear screen',
            exec => sub {
                system "clear";
                $self->restart( $cmd, $no_watch );
            },
        },
        s => {
            msg  => 'Show command',
            exec => sub {
                my $params = $self->params($cmd);
                print "\nThis terminals command:\n";
                print join ' ', $self->command($params), "\n\n";
                $self->restart( $cmd, $no_watch );
            },
        },
        r => {
            msg  => 'restart',
            exec => sub { 1 },
        },
    );

    if ( $params->{restart} ne 1 ) {
        my ($letter) = $params->{restart} =~ /^(.)/xms;
        $action{$letter} = {
            msg  => $params->{restart},
            exec => sub { exec $params->{restart}; },
        };
    }

    # show restart menu
    my $menu = "Options:\n";
    for my $letter ( sort keys %action ) {
        $menu .= "$letter - $action{$letter}{msg}\n";
    }
    print $menu;

    # get answer
    my $answer = <ARGV>;

    return if !$answer;

    chomp $answer if $answer;
    $answer ||= $params->{default} || '';

    # ask the question
    while ( !$action{$answer} ) {
        print $menu;
        print "Please choose one of " . ( join ', ', sort keys %action ) . "\n";
        $answer = <ARGV>;
        chomp $answer if $answer;
        $answer ||= $params->{default} || '';
    }

    return $action{$answer}{exec}->();
}

sub watch {
    my ( $self, $cmd ) = @_;

    my $params = $self->params($cmd);
    my @files  = $self->command(
        {
            editor => { command => undef },
            edit   => $params->{watch},
        },
    );

    my %stats;
    for my $file (@files) {
        next if !$file || !-f $file;
        $stats{$file} = stat $file;
    }

    while (1) {
        my $done = 0;
        local $SIG{INT} =
          sub { $done = $self->restart( $cmd, 1 ) ? 1 : undef; };

        sleep 1;

        for my $file (@files) {

            # return if interrupted
            return 1 if $done;

            # return if asked to quit
            return if !defined $done;

            next if !$file || !-f $file;
            my $stat = stat $file;
            return 1 if $stats{$file}->mtime ne $stat->mtime;
        }
    }

    return;
}

sub cron {
    my ( $self, $cmd ) = @_;

    my $params = $self->params($cmd);
    my $cron   = Algorithm::Cron->new(
        base    => 'local',
        crontab => $params->{cron},
    );

    while (1) {
        my $done = 0;
        local $SIG{INT} =
          sub { $done = $self->restart( $cmd, 1 ) ? 1 : undef; };

        my $next_time = $cron->next_time(time);

        #sleep $next_time - time;
        sleep 1;

        # return if interrupted
        return 1 if $done;

        # return if asked to quit
        return if !defined $done;

        if ( $params->{cron_verbose} ) {
            print {*STDERR} "\33[2K\r" . pretty_time( $next_time - time );
        }
        if ( $next_time <= time ) {
            if ( $params->{cron_verbose} ) {
                print {*STDERR} "\33[2K\r";
            }
            return 1;
        }
    }

    return;
}

sub pretty_time {
    my ($time) = @_;

    my $pretty  = '';
    my $days    = int $time / ( 24 * 60 * 60 );
    my $hours   = int $time / ( 60 * 60 ) - $days * 24;
    my $minutes = int $time / 60 - $days * 24 * 60 - $hours * 60;
    my $seconds = $time % 60;

    return "$days days $hours hours $minutes minutes $seconds seconds";
}

sub params {
    my ( $self, $cmd ) = @_;

    if ( !$cmd ) {
        warn "No \$cmd passed to params()\n", longmess();
    }

    my $config = $self->config->get;
    my $params = $config->{terminals}{$cmd} || {};

    if ( ref $params eq 'ARRAY' ) {
        $params = { command => @{$params} ? $params : '' };
    }

    if ( !$params->{command} && !$params->{edit} ) {
        $params->{command} = 'bash';
        $params->{wait}    = 0;
    }

    return merge $config->{default} || {}, $params;
}

sub command_param {
    my ( $self, $param ) = @_;

    my ($user_param) = $param =~ /^[{]:(\w+):[}]$/;

    return $param if !$user_param;

    my $value = prompt "$user_param : ";
    chomp $value;

    return $value;
}

sub command {
    my ( $self, $params, $recurse ) = @_;

    if ( !$params->{edit} ) {
        return
          ref $params->{command}
          ? map { $self->command_param($_) } @{ $params->{command} }
          : ( $params->{command} );
    }

    my $editor =
      ref $params->{editor}{command}
      ? $params->{editor}{command}
      : $self->config->get->{editor}{command};

    my @globs =
      ref $params->{edit} ? @{ $params->{edit} } : ( $params->{edit} );

    my $title = $params->{title} || $globs[0];
    my $max   = 15;
    if ( length $title > $max ) {
        $title = substr $title, ( length $title ) - $max, $max + 1;
    }

    eval { require Term::Title; }
      and Term::Title::set_titlebar($title);
    system 'tmux', 'rename-window', $title;

    my $helper_text = $self->config->get->{editor}{helper};
    my $helper;
    eval {
        if ($helper_text) {
            $helper = eval $helper_text;    ## no critic
            die "No helper generated!" if !$helper;
        }
        elsif ( $self->defaults->{verbose} ) {
            warn "No helper text";
        }
        1;
    } or do {
        warn $@;
        warn $helper_text;
    };

    my $groups = $self->config->get->{editor}{files};
    my @files  = $self->_globs2files( $groups, $helper, $recurse, @globs );

    return ( @$editor, @files );
}

sub _globs2files {
    my ( $self, $groups, $helper, $recurse, @globs ) = @_;
    my @files;
    my $count = 0;

  GLOB:
    while ( my $glob = shift @globs ) {
        last if $count++ > 30;
        my ($not_glob) = $glob =~ /^[!](.*)$/;

        if ($not_glob) {
            my %not_files = map { $_ => 1 }
              $self->_globs2files( $groups, $helper, $recurse, $not_glob );
            @files = grep { !$not_files{$_} } @files;
            next GLOB;
        }
        elsif ( $groups->{$glob} ) {
            unshift @globs, @{ $groups->{$glob} };
            next GLOB;
        }
        elsif ($helper) {
            my @g;
            eval {
                @g = $helper->( $self, $glob );
                1;
            } or do { warn $@ };

            if (@g) {
                push @files, grep { -f $_ } @g;
                unshift @globs, grep { !-f $_ } @g;
                next GLOB;
            }
        }
        if ( $recurse && -d $glob ) {
            push @files, map {
                -d $_
                  ? $self->_globs2files( $groups, $helper, $recurse, $_ )
                  : $_
            } $self->_globs2files( $groups, $helper, $recurse, "$glob/*" );
            next GLOB;
        }

        push @files, $self->_dglob($glob);
    }

    return uniq @files;
}

sub _shell_quote {
    my ($file) = @_;
    $file =~ s/([\s&;*'"])/\\$1/gxms;
    return $file;
}

sub load_env {
    my ( $self, $env_extra ) = @_;
    if ( $env_extra && ref $env_extra eq 'HASH' ) {
        for my $env ( keys %{$env_extra} ) {
            my $orig = $ENV{$env} // '';
            $ENV{$env} = $env_extra->{$env};
            $ENV{$env} =~ s/[\$]$env/$orig/xms;
        }
    }

    return;
}

sub runit {
    my ( $self, @cmd ) = @_;

    print +( join " \\\n  ", @cmd ), "\n"
      if $self->defaults->{test} || $self->defaults->{verbose};

    return if $self->defaults->{test};

    if ( @cmd > 1 ) {
        my $found = 0;
        for my $dir ( split /:/xms, $ENV{PATH} ) {
            if ( -d $dir && -x path $dir, $cmd[0] ) {
                $found = 1;
                last;
            }
        }

        if ( !$found ) {
            @cmd = ( join ' ', @cmd );
        }
    }

    $self->log( 'SYSTEM', @cmd );
    my $err = system @cmd;
    if ($err) {
        $self->log( "ERRORED", @cmd );
    }
}

sub log {
    my ( $self, @msg ) = @_;
    my $fh = path( $self->base, '.vtide', 'run.log' )->opena;
    print {$fh} '[' . localtime . "] RUN $ENV{VTIDE_TERM} " . join ' ',
      @msg . "\n";
}

sub auto_complete {
    my ($self) = @_;

    my $env   = $self->options->files->[-1];
    my @files = sort keys %{ $self->config->get->{terminals} };

    print join ' ', grep { $env ne 'run' ? /^$env/xms : 1 } @files;

    return;
}

1;

__END__

=head1 NAME

App::VTide::Command::Run - Run a terminal command

=head1 VERSION

This documentation refers to App::VTide::Command::Run version 1.0.3

=head1 SYNOPSIS

    vtide run [(-n|--name) project] [--test] terminal
    vtide run [--help|--man]

  OPTIONS:
   -n --name[=]str  The name of the terminal to run
   -T --test        Test the running of the terminal (shows the commands
                    that would be executed)
   -v --verbose     Show more verbose output.
      --help        Show this help
      --man         Show full documentation

=head1 DESCRIPTION

The C<run> command runs a terminal with what ever is configured for that
terminal. A full description of the terminal configuration can be found in
L<App::VTide::Configuration/terminals>.

=head1 SUBROUTINES/METHODS

=head2 C<run ()>

Runs the terminal command

=head2 C<restart ( $cmd )>

Checks if the terminal command should be restarted on exit (and asks if it should)

=head2 C<params ( $cmd )>

Gets the configuration for the command C<$cmd>

=head2 C<command ( $params )>

Gets the command to execute, either a simple command or an "editor" command
where the files are got from the groups

=head2 C<command_param ( $params )>

Processes any found user parameters

=head2 C<_shell_quote ( $file )>

Quote C<$file> for shell execution

=head2 C<load_env ( %env )>

Put the values of %env into the %ENV variable.

=head2 C<runit ( @cmd )>

Executes a command (with --test skipping)

=head2 C<watch ( $cmd )>

Watches files till they change then returns.

=head2 C<cron ( $cmd )>

Runs the command based on cron tab settings in params

=head2 C<pretty_time ( $time )>

Creates a mildly pretty version of the number of seconds

=head2 C<auto_complete ()>

Auto completes terminal names

=head2 C<details_sub ()>

Returns the commands details.

=head1 ATTRIBUTES

=head2 first

Track first run vs later runs (for things like waiting)

=head1 HOOKS

=head2 C<run_running ($cmd)>

Called just before execution, the command that will be executed is
passed and can be modified.

=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

=head1 DEPENDENCIES

=head1 INCOMPATIBILITIES

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2016 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut
